### PR TITLE
Add DDG::Meta::Data

### PR DESCRIPTION
**Pull request by zachthompson**
Note: This needs to be coordinated with internal changes.

DDG::Meta::Data - Mostly provides methods to instant answers so that, when interrogated, they can return metadata about themselves.  However, also includes lookup methods for cases when no IA object is available.

Some changes to dist.ini to allow container builds in travis and automatic Perl deps.

### CODE REVIEW

File: lib/DDG/Meta.pm [hunk 1]
```diff
@@ -14,6 +14,7 @@ use DDG::Meta::Block;
 use DDG::Meta::Information;
```
Comment (moollaza) [inline lib/DDG/Meta.pm]: I suppose we'll deprecate this later?

Comment (russellholt) [inline lib/DDG/Meta.pm]: yes, if we deprecate it now it will break everything

Comment (moollaza) [inline lib/DDG/Meta.pm]: :+1:

File: lib/DDG/Meta/Data.pm [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+use DDG::GoodieBundle::OpenSourceDuckDuckGo; 
+use DDG::LongtailBundle::OpenSourceDuckDuckGo; 
+use DDG::FatheadBundle::OpenSourceDuckDuckGo; 
+use JSON::XS 'decode_json';
+use Path::Class;
+use File::ShareDir 'dist_file';
+use IO::All;
+use Clone 'clone';
+
+use strict;
```
Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: you get this with `use Moo`

File: lib/DDG/Meta/Data.pm [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+my %ia_metadata;
+
+my @ia_types = qw(Spice Goodie Longtail Fathead);
+
+my %metadata_files = map { $_ => dist_file("DDG-${_}Bundle-OpenSourceDuckDuckGo", lc $_ . '/meta/metadata.json') } @ia_types;
+
+# Only build metadata once. Not in BUILD so we can call apply_keywords directly
+unless(%ia_metadata){
+
+    FILE: while (my ($type, $filename) = each %metadata_files) {
```
Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: i love this iterator

File: lib/DDG/Meta/Data.pm [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        #     "id": " " 
+        #     "signal" : " "
+        #     ....
+        # }
+        my $file_data = decode_json(io($filename)->all);
+
+        # check for decode_json_file error, returns undef
+        die "reading metadata file failed ... $filename" unless $file_data;
+
+        IA: while (my ($id, $module_data) = each %{ $file_data }) {
```
Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: is this all the same code as in core?

Comment (russellholt) [inline lib/DDG/Meta/Data.pm]: Mostly.. it's a candidate to replace it.

Comment (nilnilnil) [inline lib/DDG/Meta/Data.pm]: ah gotcha -- makes sense -- thx

File: lib/DDG/Meta/Data.pm (hunk 2 above)
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Moo is going away.

File: lib/DDG/Meta/Data.pm [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    return if $applied{$target};    
+
+    my $ias;
+    unless($ias = $self->get_ia(module => $target)){
+        warn "No metadata found for $target" if debug;
+        return;
+    }
+    # If only one id this will be false. Only a few IAs have
+    # multiple ids per module, e.g. CheatSheets
+    my $id_required = @{$ias} - 1;
```
Comment (russellholt) [inline lib/DDG/Meta/Data.pm]: do we still need this?

File: lib/DDG/Meta/Data.pm [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        unless(@m == 1){
+            die "Failed to select metadata with id $id";
+        }
+        return $m[0];
+    };
+
+    # Check for id_required *outside* of the subs so we don't incur the
+    # slight performance penalty across the board. Remember that these
+    # are method calls and that $_[0] is self
+    while(my ($k, $v) = each %{$ias->[0]}){ # must have at least one set of metadata
```
Comment (russellholt) [inline lib/DDG/Meta/Data.pm]: :+1:

File: lib/DDG/Meta/Data.pm [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        die "reading metadata file failed ... $filename" unless $file_data;
+
+        IA: while (my ($id, $module_data) = each %{ $file_data }) {
+
+            # 20150502 (zt) Can't filter like this yet as some tests depend on non-live IA metadata
+            #next unless $module_data->{status} eq 'live';
+
+            # check for bad metadata.  We need a perl_module for the by_module key
+            if($module_data->{perl_module} !~ /DDG::.+::.+/){
+                warn "Something wrong with perl_module for IA $id perl_module: $module_data->{perl_module} in $filename ...  Not fatal but the IA won't show";
```
Comment (russellholt) [inline lib/DDG/Meta/Data.pm]: skip this warn if the IA is not marked live?

Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Not a bad idea. I wanted to only load live IAs originally but we're not quite there yet.

**Comment by russellholt** [comment]
all good

**Pull request merged**